### PR TITLE
Simplify `force . delay` in UPLC

### PR DESCRIPTION
**Pull request by michaelpj**
I did this before Christmas, and I was really sure I had merged it, but apparently not...

---

This introduces a very simple simplification pass in the UPLC pipeline.
All this does is replace `force (delay t)` with `t`.

This matters because every type declaration in PIR will turn into an
immediately-instantiated type abstraction, which gets erased to a
`force/delay` pair. So this saves us ~2 AST nodes for every type
declaration, which seems like a minor but worthwhile difference.

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
    - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
    - [ ] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues
- If you changed any Purescript files:
    - [ ] `nix-shell shell.nix --run fix-purty` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-tx-plugin/src/PlutusTx/Plugin.hs [hunk 1]
```diff
@@ -335,7 +335,7 @@ runCompiler opts expr = do
     when (poDoTypecheck opts) . void $
         liftExcept $ PLC.typecheckPipeline plcTcConfig plcP
 
-    uplcP <- liftExcept $ UPLC.deBruijnProgram $ UPLC.eraseProgram plcP
+    uplcP <- liftExcept $ UPLC.deBruijnProgram $ UPLC.simplifyProgram $ UPLC.eraseProgram plcP
```
Comment (kwxm) [inline plutus-tx-plugin/src/PlutusTx/Plugin.hs:338]: If we had multiple simplifications/optimisations, would there be any point in adding something to let you turn individual things on and off?  I suppose that might be useful for testing the effects of different things, but it's probably not something you'd want to use in production.

**Review by kwxm — APPROVED** [review]
Yep, looks fine.

I'm still tempted by the idea of allowing `Force` to be applied to any term and making it a no-op on everything except `Delay` and builtins that get instantiated after all their term arguments (if that's still allowed).  That might simplify some of the runtime checks in the CEK machine.  I appreciate that some people might find it offensive though.

File: plutus-tx-plugin/src/PlutusTx/Plugin.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-tx-plugin/src/PlutusTx/Plugin.hs:338]: I think we can add this later if we need it!

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Simplify.hs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{-# LANGUAGE LambdaCase #-}
+-- | Very basic simplifications on UPLC.
+module UntypedPlutusCore.Transform.Simplify
+    ( simplifyTerm
+    , simplifyProgram
+    ) where
+
+import           UntypedPlutusCore.Core
+
+import           Control.Lens
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Simplify.hs:10]: sure

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Simplify.hs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+simplifyProgram :: Program name uni fun a -> Program name uni fun a
+simplifyProgram (Program a v t) = Program a v $ simplifyTerm t
+
+simplifyTerm :: Term name uni fun a -> Term name uni fun a
+simplifyTerm = transformOf termSubterms processTerm
```
Comment (kwxm) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Simplify.hs:16]: Hold on.  Does this get `Force (Force (Delay (Delay ...))))?  It's not immediately obvious to me if that's the case, despite having looked up `transformOf`.

Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Simplify.hs:16]: It does: there's a test to that effect! `transformOf` processes subterms first, which means that you zap the inner terms before looking at the outer ones.

Comment (kwxm) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Simplify.hs:16]: Yeah, that occurred to me just after I went out for a walk.  I'll walk before typing in future.

**Review by bezirg — APPROVED** [review]

**Pull request merged**